File: packages/dart/lib/src/location/location_generator.dart
// The location generator: one real division, or a whole location written out,
// drawn from the outline a language's dataset carries.
//
// Nothing here invents. A draw is an index into a list of divisions the country
// itself published, and every option narrows that list rather than shaping what
// is drawn from it — which is what keeps a Korean 동 inside the 구 it belongs to.

import 'dart:math';

import 'package:randino/src/constants.dart';
import 'package:randino/src/internal/generate.dart';
import 'package:randino/src/internal/parse.dart';
import 'package:randino/src/internal/utils.dart';
import 'package:randino/src/location/data/countries.dart';
import 'package:randino/src/location/data/index.dart';
import 'package:randino/src/location/data/types.dart';
import 'package:randino/src/types.dart';
import 'package:randino/src/word/data/index.dart';

/// What one draw hands back.
enum LocationForm {
  /// A single division's name, the way `randCity` returns one.
  unit,

  /// Every level down to the division written out the way the language writes a
  /// location, the way `randLocation` returns one.
  path,
}

/// The last narrowing asked of a list, and what it came to.
class _Narrowed {
  const _Narrowed(this.key, this.indexes);

  final String key;
  final List<int> indexes;
}

/// What a draw may land on, written out.
///
/// The text is worked out once, because every length and `startsWith` filter
/// reads it and a US list is thirty-two thousand of them.
abstract class _Texts {
  List<String> texts = const <String>[];
  int shortest = 0;
  int longest = 0;

  // The last narrowing asked of this list, which is the one a loop of single
  // draws asks again: filtering thirty-two thousand names is not free, and a
  // caller drawing one at a time would pay it every call.
  _Narrowed? narrowed;
}

/// The divisions one kind of draw may land on.
class _Pool extends _Texts {
  _Pool(this.data, this.language, this.form, this.withCountry, this.entries);

  final LocationLanguageData data;
  final LocationLanguage language;
  final LocationForm form;

  // Whether a location written out opens on its country. A single division's
  // name has no country to leave out.
  final bool withCountry;

  // `null` for the country, which the outline does not hold.
  final List<OutlineEntry?> entries;
}

// Parsed once per dataset, on the first draw that needs it rather than at import:
// importing the package should not cost a split of every division it ships.
final Expando<List<OutlineEntry>> _entryCache = Expando<List<OutlineEntry>>('locationEntries');

List<OutlineEntry> _entriesOf(LocationLanguageData data) =>
    _entryCache[data] ??= outline(data.outline, data.levels.length);

/// Which of the four levels an outline depth is, as an index into
/// [locationLevels].
int _rankOf(LocationLanguageData data, int depth) => locationLevels.indexOf(data.levels[depth]);

/// The divisions a draw at [level] may land on.
///
/// A single division is one at exactly that level, so a country without the
/// level has none. A location written out stops at the deepest level the country
/// has at or above the one asked for, and at any division with nothing inside it
/// at that depth: `randLocation(level: LocationLevel.city)` has to be able to
/// reach 세종특별자치시, which has no 시·군·구 and whose branch stops at the
/// region.
List<OutlineEntry?> _entriesAt(LocationLanguageData data, LocationForm form, LocationLevel level) {
  final wanted = locationLevels.indexOf(level);

  if (wanted == 0) {
    return const <OutlineEntry?>[null];
  }

  final entries = _entriesOf(data);

  if (form == LocationForm.unit) {
    return entries.where((entry) => _rankOf(data, entry.depth) == wanted).toList(growable: false);
  }

  var limit = -1;

  for (var depth = 0; depth < data.levels.length; depth += 1) {
    if (_rankOf(data, depth) <= wanted) {
      limit = depth;
    }
  }

  return entries
      .where((entry) {
        final below = entry.below;

        return entry.depth == limit || (entry.depth < limit && (below == null || below > limit));
      })
      .toList(growable: false);
}

/// One result, built fresh per draw so a caller can never reach the pool's own
/// lists.
LocationDetail _detailOf(_Pool pool, OutlineEntry? entry) {
  final data = pool.data;

  if (entry == null) {
    return LocationDetail(
      location: data.country,
      language: pool.language,
      level: LocationLevel.country,
      country: data.country,
      region: null,
      city: null,
      district: null,
    );
  }

  final named = <LocationLevel, String?>{};

  for (var depth = 0; depth < data.levels.length; depth += 1) {
    named[data.levels[depth]] = depth < entry.path.length ? entry.path[depth] : null;
  }

  final String location;

  if (pool.form == LocationForm.unit) {
    location = entry.path[entry.depth]!;
  } else {
    final parts = <String>[if (pool.withCountry) data.country, ...entry.path.nonNulls];

    location = (data.order == LocationOrder.largestFirst ? parts : parts.reversed).join(
      data.joiner,
    );
  }

  return LocationDetail(
    location: location,
    language: pool.language,
    level: data.levels[entry.depth],
    country: data.country,
    region: named[LocationLevel.region],
    city: named[LocationLevel.city],
    district: named[LocationLevel.district],
  );
}

// One pool per language, form, level and whether it opens on the country, each
// built the first time it is drawn from.
final Expando<Map<String, _Pool>> _poolCache = Expando<Map<String, _Pool>>('locationPools');

_Pool _poolOf(LocationLanguage language, LocationForm form, LocationLevel level, bool withCountry) {
  final data = locationData[language]!;
  final byKind = _poolCache[data] ??= <String, _Pool>{};
  final key = '${form.name}:${level.name}:$withCountry';
  final cached = byKind[key];

  if (cached != null) {
    return cached;
  }

  final pool = _Pool(data, language, form, withCountry, _entriesAt(data, form, level));
  final texts = List<String>.unmodifiable(
    pool.entries.map((entry) => _detailOf(pool, entry).location),
  );
  final (shortest, longest) = _spanOf(texts);

  pool
    ..texts = texts
    ..shortest = shortest
    ..longest = longest;
  byKind[key] = pool;

  return pool;
}

/// Shortest and longest of a list of texts, `(0, 0)` for none.
(int, int) _spanOf(List<String> texts) {
  var shortest = 1 << 30;
  var longest = 0;

  for (final text in texts) {
    if (text.length < shortest) shortest = text.length;
    if (text.length > longest) longest = text.length;
  }

  return texts.isEmpty ? (0, 0) : (shortest, longest);
}

/// The indexes of a pool one call may draw from once `startsWith` and the length
/// range have had their say, or `null` for all of them when neither was asked.
///
/// A range nothing in the pool fits is answered with the divisions closest to it
/// rather than with none, the way every other generator answers one — and an
/// overshoot counts half a character worse than an undershoot, because
/// `maxLength` is the bound a caller is usually holding to.
List<int>? _narrow(_Texts pool, String prefix, int? minLength, int? maxLength) {
  if (prefix.isEmpty && minLength == null && maxLength == null) {
    return null;
  }

  final key = '$prefix|$minLength|$maxLength';
  final narrowed = pool.narrowed;

  if (narrowed != null && narrowed.key == key) {
    return narrowed.indexes;
  }

  final indexes = _narrowAfresh(pool, prefix, minLength, maxLength);

  pool.narrowed = _Narrowed(key, indexes);

  return indexes;
}

List<int> _narrowAfresh(_Texts pool, String prefix, int? minLength, int? maxLength) {
  final lower = prefix.toLowerCase();
  final matching = <int>[];

  for (var index = 0; index < pool.texts.length; index += 1) {
    if (lower.isEmpty || pool.texts[index].toLowerCase().startsWith(lower)) {
      matching.add(index);
    }
  }

  if (minLength == null && maxLength == null) {
    return List<int>.unmodifiable(matching);
  }

  final bounds = lengthBounds(
    minLength,
    maxLength,
    pool.shortest,
    pool.longest,
    ceiling: randLocationLengthMax,
  );

  double missBy(int index) {
    final length = pool.texts[index].length;

    return length < bounds.min
        ? (bounds.min - length).toDouble()
        : length > bounds.max
        ? length - bounds.max + 0.5
        : 0;
  }

  var best = double.infinity;
  final closest = <int>[];

  for (final index in matching) {
    final miss = missBy(index);

    if (miss < best) {
      best = miss;
      closest
        ..clear()
        ..add(index);
    } else if (miss == best) {
      closest.add(index);
    }
  }

  return List<int>.unmodifiable(closest);
}

/// A language's pool, and the indexes of it a call may draw from.
class _Candidate<P extends _Texts> {
  const _Candidate(this.pool, this.indexes);

  final P pool;

  // `null` for every entry of the pool.
  final List<int>? indexes;
}

/// What every location generator does: a draw at one [level], in one [form].
///
/// A null [language] means every language the location generators know.
List<LocationDetail> generateLocationDetails({
  required LocationForm form,
  required LocationLevel level,
  LocationLanguage? language,
  int count = 1,
  int? minLength,
  int? maxLength,
  String? startsWith,
  bool unique = false,
  bool includeCountry = true,

  /// Where the randomness comes from: `Random.secure()` for a value nobody may
  /// predict, `Random(42)` for one that has to come out the same every run.
  Random? random,
}) {
  final prefix = resolvePrefix(startsWith);
  // A location at the country level is the country: leaving it out would leave
  // nothing, so the option only reaches the levels below it.
  final withCountry = form == LocationForm.unit || level == LocationLevel.country || includeCountry;

  // A language that cannot write the requested first character, and one with
  // nothing at the requested level, are out before a draw is made — so asking
  // every language for a 읍·면·동 draws Korean rather than spending half the
  // draws on English, which has none.
  final candidates = <_Candidate<_Pool>>[];

  for (final code in languagesWriting(language, locationLanguages, prefix)) {
    final pool = _poolOf(code, form, level, withCountry);
    final indexes =
        pool.entries.isEmpty ? const <int>[] : _narrow(pool, prefix, minLength, maxLength);

    if (indexes == null || indexes.isNotEmpty) {
      candidates.add(_Candidate(pool, indexes));
    }
  }

  if (candidates.isEmpty) {
    return <LocationDetail>[];
  }

  return withRandom(
    random,
    () => collect<LocationDetail>(
      count: count,
      unique: unique,
      startsWith: prefix,
      draw: () {
        final candidate = pick(candidates);
        final pool = candidate.pool;
        final indexes = candidate.indexes;
        final index = indexes != null ? pick(indexes) : randInt(0, pool.entries.length - 1);

        return _detailOf(pool, pool.entries[index]);
      },
      keyOf: (detail) => detail.location,
    ),
  );
}

/* --- Countries ------------------------------------------------------------ */

/// Every country's name in one language, with the code each one is known by.
class _CountryPool extends _Texts {
  _CountryPool(this.language, this.codes);

  final WordLanguage language;
  final List<String> codes;
}

// Split out of the table the first time a language is drawn from, never at
// import.
final Map<WordLanguage, _CountryPool> _countryCache = <WordLanguage, _CountryPool>{};

_CountryPool _countryPoolOf(WordLanguage language) {
  final cached = _countryCache[language];

  if (cached != null) {
    return cached;
  }

  final column = countries.languages.indexOf(language) + 1;
  final codes = <String>[];
  final texts = <String>[];

  for (final line in countries.table.split('\n')) {
    final cells = line.trim().split('|');

    if (cells.length > column) {
      codes.add(cells[0]);
      texts.add(cells[column]);
    }
  }

  final (shortest, longest) = _spanOf(texts);
  final pool = _CountryPool(language, List<String>.unmodifiable(codes));

  pool
    ..texts = List<String>.unmodifiable(texts)
    ..shortest = shortest
    ..longest = longest;
  _countryCache[language] = pool;

  return pool;
}

/// Countries in any word language: the ISO 3166-1 list, each named the way the
/// language names it.
///
/// Unlike the divisions, which only some languages have, every country has a
/// name in all nine — so this reads [language] as a word language, and a null
/// one means every word language.
List<CountryDetail> generateCountryDetails({
  WordLanguage? language,
  int count = 1,
  int? minLength,
  int? maxLength,
  String? startsWith,
  bool unique = false,

  /// Where the randomness comes from: `Random.secure()` for a value nobody may
  /// predict, `Random(42)` for one that has to come out the same every run.
  Random? random,
}) {
  final prefix = resolvePrefix(startsWith);
  final candidates = <_Candidate<_CountryPool>>[];

  for (final code in languagesWriting(language, wordLanguages, prefix)) {
    final pool = _countryPoolOf(code);
    final indexes = _narrow(pool, prefix, minLength, maxLength);

    if (indexes == null || indexes.isNotEmpty) {
      candidates.add(_Candidate(pool, indexes));
    }
  }

  if (candidates.isEmpty) {
    return <CountryDetail>[];
  }

  return withRandom(
    random,
    () => collect<CountryDetail>(
      count: count,
      unique: unique,
      startsWith: prefix,
      draw: () {
        final candidate = pick(candidates);
        final pool = candidate.pool;
        final indexes = candidate.indexes;
        final index = indexes != null ? pick(indexes) : randInt(0, pool.codes.length - 1);

        return CountryDetail(
          country: pool.texts[index],
          code: pool.codes[index],
          language: pool.language,
        );
      },
      keyOf: (detail) => detail.country,
    ),
  );
}

File: packages/javascript/lib/location/locationGenerator.ts
// The location generator: one real division, or a whole location written out,
// drawn from the outline a language's dataset carries.
//
// Nothing here invents. A draw is an index into a list of divisions the country
// itself published, and every option narrows that list rather than shaping what
// is drawn from it — which is what keeps a Korean 동 inside the 구 it belongs to.

import {
	collect,
	languagesWriting,
	lengthBounds,
	resolveLength,
	resolvePrefix,
	resolveRandom
} from '../_internal/generate.js';
import { outline } from '../_internal/parse.js';
import type { OutlineEntry } from '../_internal/parse.js';
import { pick, randInt, withRandom } from '../_internal/utils.js';
import { RAND_LOCATION_LENGTH_MAX } from '../constants.js';
import type {
	CountryDetail,
	LocationDetail,
	LocationLanguage,
	LocationLevel,
	RandCountryOptions,
	RandLocationUnitOptions,
	WordLanguage
} from '../_types/global.js';
import { WORD_LANGUAGES, resolveWordLanguage } from '../word/data/index.js';
import { COUNTRIES } from './data/countries.js';
import {
	LOCATION_DATA,
	LOCATION_LANGUAGES,
	LOCATION_LEVELS,
	resolveLocationLanguage
} from './data/index.js';
import type { LocationLanguageData } from './data/types.js';

/**
 * What one draw hands back: a single division's name (`randCity`), or every
 * level down to it written out the way the language writes a location
 * (`randLocation`).
 */
export type LocationForm = 'unit' | 'path';

/**
 * What a draw may land on, written out. The text is worked out once, because
 * every length and `startsWith` filter reads it and a US list is thirty-two
 * thousand of them.
 */
type Texts = {
	texts: readonly string[];
	shortest: number;
	longest: number;
	// The last narrowing asked of this list, which is the one a loop of single
	// draws asks again: filtering thirty-two thousand names costs a millisecond,
	// and a caller drawing one at a time would pay it every call.
	narrowed?: { key: string; indexes: readonly number[] };
};

/** The divisions one kind of draw may land on. */
type Pool = Texts & {
	data: LocationLanguageData;
	language: LocationLanguage;
	form: LocationForm;
	// Whether a location written out opens on its country. A single division's
	// name has no country to leave out.
	withCountry: boolean;
	// `null` for the country, which the outline does not hold.
	entries: readonly (OutlineEntry | null)[];
};

// Parsed once per dataset, on the first draw that needs it rather than at import:
// importing the package should not cost a split of every division it ships.
const entryCache = new WeakMap<LocationLanguageData, readonly OutlineEntry[]>();

function entriesOf(data: LocationLanguageData): readonly OutlineEntry[] {
	let entries = entryCache.get(data);

	if (!entries) {
		entries = outline(data.outline, data.levels.length);
		entryCache.set(data, entries);
	}

	return entries;
}

/** Which of the four levels an outline depth is, as an index into `LOCATION_LEVELS`. */
function rankOf(data: LocationLanguageData, depth: number): number {
	return LOCATION_LEVELS.indexOf(data.levels[depth]);
}

/**
 * The divisions a draw at `level` may land on.
 *
 * A single division is one at exactly that level, so a country without the level
 * has none. A location written out stops at the deepest level the country has at
 * or above the one asked for, and at any division with nothing inside it at that
 * depth: `randLocation({ level: 'city' })` has to be able to reach
 * 세종특별자치시, which has no 시·군·구 and whose branch stops at the region.
 */
function entriesAt(
	data: LocationLanguageData,
	form: LocationForm,
	level: LocationLevel
): readonly (OutlineEntry | null)[] {
	const wanted = LOCATION_LEVELS.indexOf(level);

	if (wanted === 0) {
		return [null];
	}

	const entries = entriesOf(data);

	if (form === 'unit') {
		return entries.filter((entry) => rankOf(data, entry.depth) === wanted);
	}

	let limit = -1;

	data.levels.forEach((_, depth) => {
		if (rankOf(data, depth) <= wanted) {
			limit = depth;
		}
	});

	return entries.filter(
		(entry) =>
			entry.depth === limit ||
			(entry.depth < limit && (entry.below === null || entry.below > limit))
	);
}

/** One result, built fresh per draw so a caller can never reach the pool's own arrays. */
function detailOf(pool: Pool, entry: OutlineEntry | null): LocationDetail {
	const { data } = pool;
	const named: Record<'region' | 'city' | 'district', string | null> = {
		region: null,
		city: null,
		district: null
	};

	if (!entry) {
		return {
			location: data.country,
			language: pool.language,
			level: 'country',
			country: data.country,
			...named
		};
	}

	data.levels.forEach((level, depth) => {
		named[level] = entry.path[depth] ?? null;
	});

	const parts = [pool.withCountry ? data.country : null, ...entry.path].filter(
		(part): part is string => part !== null
	);
	const name = entry.path[entry.depth] as string;

	return {
		location:
			pool.form === 'unit'
				? name
				: (data.order === 'largest-first' ? parts : parts.reverse()).join(data.joiner),
		language: pool.language,
		level: data.levels[entry.depth],
		country: data.country,
		...named
	};
}

// One pool per language, form, level and whether it opens on the country, each
// built the first time it is drawn from.
const poolCache = new WeakMap<LocationLanguageData, Map<string, Pool>>();

function poolOf(
	language: LocationLanguage,
	form: LocationForm,
	level: LocationLevel,
	withCountry: boolean
): Pool {
	const data = LOCATION_DATA[language];
	let byKind = poolCache.get(data);

	if (!byKind) {
		byKind = new Map();
		poolCache.set(data, byKind);
	}

	const key = `${form}:${level}:${withCountry}`;
	const cached = byKind.get(key);

	if (cached) {
		return cached;
	}

	const entries = entriesAt(data, form, level);
	const pool: Pool = {
		data,
		language,
		form,
		withCountry,
		entries,
		texts: [],
		shortest: 0,
		longest: 0
	};
	const texts = entries.map((entry) => detailOf(pool, entry).location);

	pool.texts = texts;
	[pool.shortest, pool.longest] = spanOf(texts);
	byKind.set(key, pool);

	return pool;
}

/** Shortest and longest of a list of texts, `[0, 0]` for none. */
function spanOf(texts: readonly string[]): [number, number] {
	let shortest = Infinity;
	let longest = 0;

	// A loop rather than `Math.min(...lengths)`: spreading thirty-two thousand
	// arguments is close enough to the engine's limit to fail on the next dataset.
	for (const text of texts) {
		shortest = Math.min(shortest, text.length);
		longest = Math.max(longest, text.length);
	}

	return texts.length ? [shortest, longest] : [0, 0];
}

/**
 * The indexes of a pool one call may draw from once `startsWith` and the length
 * range have had their say, or `null` for all of them when neither was asked.
 *
 * A range nothing in the pool fits is answered with the divisions closest to it
 * rather than with none, the way every other generator answers one — and an
 * overshoot counts half a character worse than an undershoot, because
 * `maxLength` is the bound a caller is usually holding to.
 */
function narrow(
	pool: Texts,
	prefix: string,
	minLength: number | undefined,
	maxLength: number | undefined
): readonly number[] | null {
	if (!prefix && minLength === undefined && maxLength === undefined) {
		return null;
	}

	const key = `${prefix}|${minLength}|${maxLength}`;

	if (pool.narrowed?.key === key) {
		return pool.narrowed.indexes;
	}

	const indexes = narrowAfresh(pool, prefix, minLength, maxLength);

	pool.narrowed = { key, indexes };

	return indexes;
}

function narrowAfresh(
	pool: Texts,
	prefix: string,
	minLength: number | undefined,
	maxLength: number | undefined
): readonly number[] {
	const lower = prefix.toLowerCase();
	const matching: number[] = [];

	pool.texts.forEach((text, index) => {
		if (!lower || text.toLowerCase().startsWith(lower)) {
			matching.push(index);
		}
	});

	if (minLength === undefined && maxLength === undefined) {
		return matching;
	}

	const [min, max] = lengthBounds(
		minLength,
		maxLength,
		pool.shortest,
		pool.longest,
		RAND_LOCATION_LENGTH_MAX
	);
	const missBy = (index: number) => {
		const length = pool.texts[index].length;

		return length < min ? min - length : length > max ? length - max + 0.5 : 0;
	};
	let best = Infinity;
	let closest: number[] = [];

	for (const index of matching) {
		const miss = missBy(index);

		if (miss < best) {
			best = miss;
			closest = [index];
		} else if (miss === best) {
			closest.push(index);
		}
	}

	return closest;
}

export function generateLocationDetails(
	form: LocationForm,
	level: LocationLevel,
	options: RandLocationUnitOptions = {},
	includeCountry = true
): LocationDetail[] {
	// A location at the country level is the country: leaving it out would leave
	// nothing, so the option only reaches the levels below it.
	const withCountry = form === 'unit' || level === 'country' || includeCountry;
	const language = resolveLocationLanguage(options.language);
	const prefix = resolvePrefix(options.startsWith);
	const minLength = resolveLength(options.minLength);
	const maxLength = resolveLength(options.maxLength);

	// A language that cannot write the requested first character, and one with
	// nothing at the requested level, are out before a draw is made — so asking
	// every language for a 읍·면·동 draws Korean rather than spending half the
	// draws on English, which has none.
	const candidates = languagesWriting(language, LOCATION_LANGUAGES, prefix).flatMap((code) => {
		const pool = poolOf(code, form, level, withCountry);
		const indexes = pool.entries.length ? narrow(pool, prefix, minLength, maxLength) : [];

		return indexes === null || indexes.length ? [{ pool, indexes }] : [];
	});

	if (!candidates.length) {
		return [];
	}

	return withRandom(resolveRandom(options.random), () =>
		collect(
			options,
			() => {
				const { pool, indexes } = pick(candidates);
				const index = indexes ? pick(indexes) : randInt(0, pool.entries.length - 1);

				return detailOf(pool, pool.entries[index]);
			},
			(detail) => detail.location
		)
	);
}

/** What every location generator does: a draw at one level, in one form. */
export function drawLocation(
	form: LocationForm,
	level: LocationLevel,
	options: RandLocationUnitOptions & { output?: 'value' | 'detail' },
	includeCountry = true
): string[] | LocationDetail[] {
	const details = generateLocationDetails(form, level, options, includeCountry);

	return options.output === 'detail' ? details : details.map((detail) => detail.location);
}

/* --- Countries ------------------------------------------------------------ */

/** Every country's name in one language, with the code each one is known by. */
type CountryPool = Texts & { language: WordLanguage; codes: readonly string[] };

// Split out of the table the first time a language is drawn from, never at import.
const countryCache = new Map<WordLanguage, CountryPool>();

function countryPoolOf(language: WordLanguage): CountryPool {
	const cached = countryCache.get(language);

	if (cached) {
		return cached;
	}

	const column = COUNTRIES.languages.indexOf(language) + 1;
	const codes: string[] = [];
	const texts: string[] = [];

	for (const line of COUNTRIES.table.split('\n')) {
		const cells = line.trim().split('|');

		if (cells.length > column) {
			codes.push(cells[0]);
			texts.push(cells[column]);
		}
	}

	const [shortest, longest] = spanOf(texts);
	const pool: CountryPool = { language, codes, texts, shortest, longest };

	countryCache.set(language, pool);

	return pool;
}

/**
 * Countries in any word language: the ISO 3166-1 list, each named the way the
 * language names it. Unlike the divisions, which only some languages have, every
 * country has a name in all nine — so this reads `language` as a word language.
 */
export function generateCountryDetails(options: RandCountryOptions = {}): CountryDetail[] {
	const language = resolveWordLanguage(options.language);
	const prefix = resolvePrefix(options.startsWith);
	const minLength = resolveLength(options.minLength);
	const maxLength = resolveLength(options.maxLength);
	const candidates = languagesWriting(language, WORD_LANGUAGES, prefix).flatMap((code) => {
		const pool = countryPoolOf(code);
		const indexes = narrow(pool, prefix, minLength, maxLength);

		return indexes === null || indexes.length ? [{ pool, indexes }] : [];
	});

	if (!candidates.length) {
		return [];
	}

	return withRandom(resolveRandom(options.random), () =>
		collect(
			options,
			() => {
				const { pool, indexes } = pick(candidates);
				const index = indexes ? pick(indexes) : randInt(0, pool.codes.length - 1);

				return { country: pool.texts[index], code: pool.codes[index], language: pool.language };
			},
			(detail) => detail.country
		)
	);
}
